### PR TITLE
Fix categorical primitives distribution argument

### PR DESCRIPTION
**Pull request by ahtsan**
Without providing the function argument `probs`, the default argument is `logits`
which is not the vanilla probability values.

### CODE REVIEW

**Comment by ahtsan** [comment]
I can't fully explain yet, but if we are passing the output from MLP to the Categorical distribution as `probs` instead of `logits`, it learns better.

**Comment by ahtsan** [comment]
Are tf.distribution part of 2020.03 release?

**Comment by ryanjulian** [comment]
They aren't blocking the release, no.

How much better does it learn? How do you know?

```
logits: An N-D Tensor, N >= 1, representing the unnormalized log probabilities of a set of Categorical distributions. The first N - 1 dimensions index into a batch of independent distributions and the last dimension represents a vector of logits for each class. Only one of logits or probs should be passed in.
probs: An N-D Tensor, N >= 1, representing the probabilities of a set of Categorical distributions. The first N - 1 dimensions index into a batch of independent distributions and the last dimension represents a vector of probabilities for each class. Only one of logits or probs should be passed in.
```

Is your MLP output log-prob or absolute? are they guaranteed normalized?

**Comment by ahtsan** [comment]
I think it depends on our interpretation? MLP is just an estimated function that we train to estimate whatever values. Would it be possible the values are too large for log-prob?

**Comment by ryanjulian** [comment]
My main concern is that the documentation above implies that probs must be normalized (i.e. sum to 1).

is it just as effective if you normalize the output of the MLP before sending it to prob?

**Comment by ahtsan** [comment]
It is. Updated.

File: src/garage/tf/models/categorical_mlp_model.py [hunk 1]
```diff
@@ -72,4 +72,5 @@ def _build(self, state_input, name=None):
 
         """
         prob = super()._build(state_input, name=name)
-        return tfp.distributions.OneHotCategorical(prob)
+        prob = prob / tf.math.reduce_sum(prob)
```
Comment (ryanjulian) [inline src/garage/tf/models/categorical_mlp_model.py:75]: i think you should probably use https://www.tensorflow.org/api_docs/python/tf/norm `ord=1` rather than a division operator.

your implementation is dangerous if `prob` is very small or zero. i think tf.norm should handle this? but maybe not.

Comment (ahtsan) [reply src/garage/tf/models/categorical_mlp_model.py:75]: How about we manually set the `output_nonlinearity` of the MLP to softmax?

Comment (ahtsan) [reply src/garage/tf/models/categorical_mlp_model.py:75]: I realized we were already using softmax so the output of the MLP should already be normalized and sum to 1?

Comment (ahtsan) [reply src/garage/tf/models/categorical_mlp_model.py:75]: I can add a test

Comment (ryanjulian) [reply src/garage/tf/models/categorical_mlp_model.py]: softmax should sum to 1

File: src/garage/tf/models/categorical_gru_model.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def __init__(self,
                  hidden_b_init=tf.zeros_initializer(),
                  recurrent_nonlinearity=tf.nn.sigmoid,
                  recurrent_w_init=tf.initializers.glorot_uniform(),
-                 output_nonlinearity=None,
```
Comment (ryanjulian) [inline src/garage/tf/models/categorical_gru_model.py:63]: if this is necessary for the correct operation of the model, it should not be an option.

Alternatively, you could rename this something like `output_normalization_fn`, but I recommend just removing it.

File: src/garage/tf/models/categorical_mlp_model.py [hunk 3]
```diff
@@ -47,10 +47,10 @@ def __init__(self,
                  output_dim,
                  name=None,
                  hidden_sizes=(32, 32),
-                 hidden_nonlinearity=tf.nn.relu,
+                 hidden_nonlinearity=tf.nn.tanh,
                  hidden_w_init=tf.initializers.glorot_uniform(),
                  hidden_b_init=tf.zeros_initializer(),
-                 output_nonlinearity=None,
+                 output_nonlinearity=tf.nn.softmax,
```
Comment (ryanjulian) [inline src/garage/tf/models/categorical_mlp_model.py:53]: see my comment above.

File: src/garage/tf/models/categorical_gru_model.py (hunk 2 above)
Comment (ahtsan) [reply src/garage/tf/models/categorical_gru_model.py:63]: Right, we should have enforced it for categorical primitives. Will fix.

Comment (ryanjulian) [reply src/garage/tf/models/categorical_gru_model.py:63]: there's also no reason you can't have an output non-linearity followed by your normalizer. that seems like another way out of this problem

Comment (ahtsan) [reply src/garage/tf/models/categorical_gru_model.py:63]: Indeed, how about applying softmax inside the model, while making `output_nonlineaity=None` by default? Then we can still have any output non-linearity as we want.

Comment (ryanjulian) [reply src/garage/tf/models/categorical_gru_model.py:63]: lgtm

**Comment by ahtsan** [comment]
@krzentner @naeioi PTAL

File: src/garage/tf/policies/categorical_gru_policy2.py [hunk 4]
```diff
@@ -67,7 +67,7 @@ def __init__(self,
                  hidden_b_init=tf.zeros_initializer(),
                  recurrent_nonlinearity=tf.nn.sigmoid,
                  recurrent_w_init=tf.initializers.glorot_uniform(),
-                 output_nonlinearity=tf.nn.softmax,
+                 output_nonlinearity=None,
```
Comment (krzentner) [inline src/garage/tf/policies/categorical_gru_policy2.py:70]: Is this because it's already done by a submodel? If so, does it ever makes sense to set this parameter? If not, should we delete it?

Comment (ahtsan) [reply src/garage/tf/policies/categorical_gru_policy2.py:70]: Submodel already applied a softmax layer. This output_nonlinearity allows users to do whatever they want.
Currently all primitives are sharing the same protocol, i.e. hidden_w, hidden_b, hidden_nonlinearity, output_w, output_b, output_nonlinearity.
We can also decide to remove some of them if we think some primitives do not need them, but I think that's not what we should do at this moment.

Comment (krzentner) [reply src/garage/tf/policies/categorical_gru_policy2.py:70]: Okay, sounds good.

**Comment by ahtsan** [comment]
@ryanjulian I don't think it needs backport? It wasn't in 2019.10 release.

**Comment by ryanjulian** [comment]
@ahtsan you're right -- removed it
